src/test/scala/chiselTests/naming/IdentifierProposerSpec.scala

```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chiselTests
+package naming
+
+import chisel3._
+import chisel3.naming.IdentifierProposer
+
+case class Blah(i: Int)
+class Foo(val i: Int)
+object Foo {
+  implicit val fooProposer = new IdentifierProposer[Foo] {
+    def propose(foo: Foo): String = s"Foo(${foo.i})"
+  }
```

Comment:
This appears to be unused, related to my comments about the "Typeclass-y API not being implemented [yet]".

Reply:
Are you sure it's being used? Because `Blah` is getting the exact same style of name below. I'd like to see an example that does an actually interesting proposal to prove it's being used. I'm skeptical because I had an extremely difficult time getting implicit resolution to work when generating the methods in the compiler plugin. I suspect it _is_ possible, but I don't see it here yet. And even if this case works (because it's defined in implicit scope), we need to show that the implicit resolution follows the rules, try putting the type class definition in a different object and importing it into lexical scope--that was the case I could not get to work.

Reply:
Yeah it works, any other cases I should test?

Reply:
I realize I commented in the wrong place, it of course works when you call `IdentifierProposer.getProposal`, but it does **not** work for actually customizing the name of a Module, the test that I'm concerned about it:
```scala
class MyModule(i: Int) extends Module
object MyModule {
  implicit val proposer = new IdentifierProposer[MyModule] {
    def propose(x: MyModule): String = "SomeOtherName"                                                                                                                                                         
  }
}

Module(new MyModule(0)).definitionIdentifier // This should be "SomeOtherName" but is "MyModule_0"
```
I want to do the typeclass API, but if it doesn't work yet, I'm not sure that we should expose it as public.

Reply:
After offline conversation w/ @jackkoenig , we decided the typeclass customization would be too onerous to do because you'd need to pepper `: IdentiferProposer` everywhere, so instead we will just use inheritance.